bundles/org.eclipse.ui.workbench/eclipseui/org/eclipse/ui/internal/e4/compatibility/CompatibilityPart.java

```diff
@@ -420,7 +421,12 @@ public void create() {
 	@PreDestroy
 	void destroy() {
 		if (!alreadyDisposed) {
-			invalidate();
+			Display display = Display.getDefault();
+			if (display != null && !display.isDisposed() && Display.getCurrent() != display) {
+				display.syncExec(this::invalidate);
+			} else {
+				invalidate();
+			}
```

Comment:
E4 Has the `UISynchronize` service for this purposes:


```suggestion
void destroy(UISynchronize sync) {
if (!alreadyDisposed) {
sync.syncExec ...
```

I also wonder if not the caller is to blame not calling from the UI thread, so maybe this should better be fixed than on the client side.

Reply:
yes, thanks for the newer way. updated.